### ISSUE TITLE
False positive "Cast is redundant"

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#37041]
VSF_TYPE_MARKDOWNVisual Studio says that cast in list. Add((Func&lt;B&gt;) Foo); is redundant in the following code (but it is not redundant, removing the cast changes the output of the program).
```cs
using System;
using System.Collections.Generic;

namespace RedundantCast
{
    class Program
    {
        class A { }
        class B : A { }

        B Foo() { return null; }

        static void Main()
        {
            (new Program()). Run();
        }

        void Run()
        {
            var list = new List<Func<A>>();
            list. Add((Func<B>) Foo);
            switch (list[0])
            {
                case Func<B> value: Console.WriteLine("B"); break;
                case Func<A> value: Console.WriteLine("A"); break;
            }
        }
    }
}
```
_This issue has been moved from https://developercommunity.visualstudio.com/content/problem/635779/false-positive-cast-is-redundant.html
VSTS ticketId: 943413_
_These are the original issue comments:_

Visual Studio Feedback System on 7/8/2019, 02:11 AM (8 hours ago): <markdown><p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>
</markdown>
_These are the original issue solutions:_
(no solutions)
